### PR TITLE
WIP: Modify metronome GC to use object scanners

### PR DESCRIPTION
**Pull request by jason-hall**
Signed-off-by: Jason Hall <jasonhal@ca.ibm.com>

### CODE REVIEW

**Comment by jason-hall** [comment]
@dmitripivkine Could I get some feedback on the direction this approach is headed?

**Comment by dmitripivkine** [comment]
@amicic FYI

**Comment by jason-hall** [comment]
Related omr change: https://github.com/jason-hall/omr/commit/c857bfd239c1a5b2da6f123a7e2155cbbf4e65ce

File: runtime/gc_glue_java/MarkingDelegate.hpp [hunk 1]
```diff
@@ -112,63 +119,79 @@ class MM_MarkingDelegate
...
-		}
-		case GC_ObjectModel::SCAN_PRIMITIVE_ARRAY_OBJECT:
-			/* nothing to do */
-			*sizeToDo = 0;
-			break;
-		default:
-			Assert_MM_unreachable();
-		}
+#if defined (J9VM_GC_REALTIME)
+		if (IS_ITEM_ARRAYLET((uintptr_t)objectPtr)) {
```
Comment (amicic) [inline runtime/gc_glue_java/MarkingDelegate.hpp]: I don't like penalizing base objects scanner (used in non-realtime). Let's leave this check in incrementalCompleteScan, for now. Unfortunate that means we won't be able to share incrementalCompleteScan and base completeScan, but it's ok for now.

File: runtime/gc_realtime/RealtimeMarkingScheme.cpp [hunk 2]
```diff
@@ -215,3 +209,89 @@ MM_RealtimeMarkingScheme::incrementalCompleteScan(MM_EnvironmentRealtime *env, u
...
+	GC_ObjectScanner *objectScanner = _delegate.getObjectScanner(env, (omrobjectptr_t)item, objectScannerState, SCAN_REASON_PACKET, &sizeToDo);
+	if (NULL != objectScanner) {
+		bool isLeafSlot = false;
+		GC_SlotObject *slotObject;
+		while (NULL != (slotObject = objectScanner->getNextSlot())) {
+			fixupForwardedSlot(slotObject);
+			inlineMarkObjectNoCheck(env, slotObject->readReferenceFromSlot(), isLeafSlot);
+			scannedPointers += 1;
+		}
+	}
```
Comment (amicic) [inline runtime/gc_realtime/RealtimeMarkingScheme.cpp]: Let's call base scanObject - make it inline on base side if possible. Besides, isLeafSlot is not properly setup.

Comment (amicic) [reply runtime/gc_realtime/RealtimeMarkingScheme.cpp]: Hm... I guess, you need sizeToDo and scannedPointers. Let me think more.

Comment (amicic) [reply runtime/gc_realtime/RealtimeMarkingScheme.cpp]: MM_EnvironmentRealtime::_scannedPointerFields  does not see to be used, so you can get rid of it and related getter/setter.

Comment (amicic) [reply runtime/gc_realtime/RealtimeMarkingScheme.cpp]: Same for _scannedBytes and _scannedObjects.

Hence this whole scanObject becomes identical to base scanObject and you can get rid of this one altogether.

Comment (amicic) [reply runtime/gc_realtime/RealtimeMarkingScheme.cpp]: Argh, they still differ in return value: base one returns _bytes_ scanned, while realtime one expects scannedPointers to be returned....

Comment (amicic) [reply runtime/gc_realtime/RealtimeMarkingScheme.cpp]: You can extend env->_markStats and in base scanObject  to maintain _slotsScanned (similar to _objectsScanned). Then in Realtime completeScan loop don't rely on scanObject  return value but on env->_markStats._slotsScanned

File: runtime/gc_glue_java/MarkingDelegate.cpp [hunk 3]
```diff
@@ -600,62 +601,7 @@ MM_MarkingDelegate::setupReferenceObjectScanner(MM_EnvironmentBase *env, omrobje
...
-			/* immediately make the next chunk available for another thread to start processing */
-			uintptr_t nextIndex = startIndex + slotsToScan;
-			Assert_MM_true(nextIndex < sizeInElements);
-			void *element1 = (void *)objectPtr;
-			void *element2 = (void *)((nextIndex << PACKET_ARRAY_SPLIT_SHIFT) | PACKET_ARRAY_SPLIT_TAG);
-			Assert_MM_true(nextIndex == (((uintptr_t)element2) >> PACKET_ARRAY_SPLIT_SHIFT));
-			env->_workStack.push(env, element1, element2);
-			env->_workStack.flushOutputPacket(env);
-			MM_MarkJavaStats *markJavaStats = &(env->getGCEnvironment()->_markJavaStats);
-			markJavaStats->splitArraysProcessed += 1;
```
Comment (amicic) [inline runtime/gc_glue_java/MarkingDelegate.cpp]: It's good that you are moving this to OMR, but you'll have to move splitArraysProcessed and splitArraysAmount  to OMR Mark Stats.

File: runtime/gc_glue_java/MetronomeDelegate.hpp [hunk 4]
```diff
@@ -197,305 +197,6 @@ class MM_MetronomeDelegate : public MM_BaseNonVirtual
...
-						env->getWorkStack()->push(env, (void *)ARRAYLET_TO_ITEM(startScanPtr));
-					}
-				}
-			}
-		}
-
-		/* check for yield if we've actually scanned a leaf */
-		if (0 != pointerFields) {
-			_scheduler->condYieldFromGC(env);
-		}
```
Comment (amicic) [inline runtime/gc_glue_java/MetronomeDelegate.hpp]: It's hard to follow if this method changed when you moved to RealtimeMarking scheme, but as I minimum, this yield check seems to be lost.

File: runtime/gc_glue_java/MarkingDelegate.hpp [hunk 5]
```diff
@@ -112,63 +119,79 @@ class MM_MarkingDelegate
...
+				objectScanner = GC_MixedObjectScanner::newInstance(env, objectPtr, scannerSpace, 0);
+				*sizeToDo = sizeof(fomrobject_t) + ((GC_MixedObjectScanner *)objectScanner)->getBytesRemaining();
+				break;
+			case GC_ObjectModel::SCAN_POINTER_ARRAY_OBJECT:
+			{
+				uintptr_t slotsToDo = 0;
+				uintptr_t flags = 0;
+				fomrobject_t *basePtr = 0;
+				uintptr_t sizeInElements = 0;
+				uintptr_t startIndex = setupPointerArrayScanner(env, objectPtr, reason, sizeToDo, &slotsToDo, &sizeInElements, &basePtr, &flags);
```
Comment (amicic) [inline runtime/gc_glue_java/MarkingDelegate.hpp]: I'd slightly simplify the signature of setupPointerArrayScanner so that instead of passing triplet sizeToDo, &slotsToDo, &sizeInElements, pass only the last one (and call it sizeInElementsToDo), and calculate the first two here.

File: runtime/gc_glue_java/PointerArrayObjectScanner.hpp [hunk 6]
```diff
@@ -84,14 +84,10 @@ class GC_PointerArrayObjectScanner : public GC_IndexableObjectScanner
 	 * @param[in] startIndex The index of the first element to scan
 	 */
 	MMINLINE static GC_PointerArrayObjectScanner *
-	newInstance(MM_EnvironmentBase *env, omrobjectptr_t objectPtr, void *allocSpace, uintptr_t flags, uintptr_t splitAmount, uintptr_t startIndex = 0)
+	newInstance(MM_EnvironmentBase *env, omrobjectptr_t objectPtr, fomrobject_t *basePtr, void *allocSpace, uintptr_t flags, uintptr_t splitAmount, uintptr_t sizeInElements,  uintptr_t startIndex = 0)
```
Comment (amicic) [inline runtime/gc_glue_java/PointerArrayObjectScanner.hpp]: I have a feeling that we don't need sizeInElements.

If split flag is not set, splitAmount and sizeInElements should have the same value.

Comment (amicic) [reply runtime/gc_glue_java/PointerArrayObjectScanner.hpp]: Or I should say, we don't need splitAmount anymore - just rename splitAmount to sizeInElements, and document that if:
1) splitting is used, it's the total size of array
2) splitting is not used, it's the rest to be scanned.

File: runtime/gc_realtime/RealtimeMarkingScheme.cpp [hunk 7]
```diff
@@ -215,3 +209,98 @@ MM_RealtimeMarkingScheme::incrementalCompleteScan(MM_EnvironmentRealtime *env, u
 	}
 }
 
+uintptr_t
+MM_RealtimeMarkingScheme::scanObject(MM_EnvironmentRealtime *env, uintptr_t item, GC_ObjectScannerState *objectScannerState)
+{
+	uintptr_t scannedPointers = 0;
+	uintptr_t sizeToDo = UDATA_MAX;
+	GC_ObjectScanner *objectScanner = nullptr;
+	if (IS_ITEM_ARRAYLET((uintptr_t)objectPtr)) {
```
Comment (jason-hall) [inline runtime/gc_realtime/RealtimeMarkingScheme.cpp]: I put the IS_ITEM_ARRAYLET check into scanObject instead of incrementalCompleteScan because that's where the created scanner is used. Creating the scanner in incrementalScan() and trying to use base scanObject() maybe doesn't seem right?

File: runtime/gc_realtime/RealtimeMarkingScheme.cpp [hunk 8]
```diff
@@ -215,3 +209,98 @@ MM_RealtimeMarkingScheme::incrementalCompleteScan(MM_EnvironmentRealtime *env, u
...
+		sizeToDo =  _omrVM->arrayletLeafSize;
+		objectScanner = GC_PointerArrayObjectScanner::newInstance(env, objectPtr, ITEM_TO_ARRAYLET(objectPtr), scannerSpace, GC_ObjectScanner::indexableObjectNoSplit, sizeInElements, sizeInElements);
+	} else {
+		objectScanner = _delegate.getObjectScanner(env, (omrobjectptr_t)item, objectScannerState, SCAN_REASON_PACKET, &sizeToDo);
+	}
+	if (NULL != objectScanner) {
+		bool isLeafSlot = false;
+		GC_SlotObject *slotObject;
+#if defined(OMR_GC_LEAF_BITS)
+		while (NULL != (slotObject = objectScanner->getNextSlot(&isLeafSlot))) {
```
Comment (jason-hall) [inline runtime/gc_realtime/RealtimeMarkingScheme.cpp]: Fixed isLeafSlot check

File: runtime/gc_stats/MarkJavaStats.hpp [hunk 9]
```diff
@@ -53,11 +53,6 @@ class MM_MarkJavaStats : public MM_Base {
 	UDATA _stringConstantsCleared; /**< The number of string constants that have been cleared during marking */
 	UDATA _stringConstantsCandidates; /**< The number of string constants that have been visited in string table during marking */
 
-#if defined(J9MODRON_TGC_PARALLEL_STATISTICS)
-	UDATA splitArraysProcessed; /**< The number of array chunks (not counting parts smaller than the split size) processed by this thread */
```
Comment (jason-hall) [inline runtime/gc_stats/MarkJavaStats.hpp]: splitArraysProcessed moving to omr stats

File: runtime/gc_glue_java/MetronomeDelegate.hpp (hunk 4 above)
Comment (jason-hall) [reply runtime/gc_glue_java/MetronomeDelegate.hpp]: condYieldFromGC shouldn't actually be lost since incrementalCompleteScan() regularly does it as objects are scanned? 
 I think this one was just for the leaf slot scan and is now done with a scanner and condYieldFromGC in incrementalCompleteScan()? I'm not sure of a better way

File: runtime/gc_realtime/RealtimeMarkingScheme.cpp [hunk 10]
```diff
@@ -215,3 +209,98 @@ MM_RealtimeMarkingScheme::incrementalCompleteScan(MM_EnvironmentRealtime *env, u
...
+	env->addScannedBytes(sizeToDo);
+
+	env->addScannedPointerFields(scannedPointers);
+	env->incScannedObjects();
+
+	return scannedPointers;
+}
+
+uintptr_t
+MM_RealtimeMarkingScheme::setupPointerArrayScanner(MM_EnvironmentBase *env, omrobjectptr_t objectPtr, MM_MarkingSchemeScanReason reason, uintptr_t *sizeInElementsToDo, fomrobject_t **basePtr, uintptr_t *flags)
```
Comment (jason-hall) [inline runtime/gc_realtime/RealtimeMarkingScheme.cpp]: Fixed up excess parameters to setupPointerArrayScanner

**Comment by jason-hall** [comment]
@amicic Rebased to latest. Edited commit/PR comment. I think this work was complete last we looked at it.

**Comment by genie-openj9** [comment]
Can one of the admins verify this patch?
